### PR TITLE
[TRTLLM-6409][feat] Enable guided decoding with speculative decoding (part 1: two-model engine)

### PR DESCRIPTION
**Pull request by syuoni**
## Summary by CodeRabbit

* **New Features**
  * Added support for rollback and termination status in grammar matcher components, enhancing guided decoding capabilities.
  * Enabled guided decoding to work alongside speculative decoding, supporting multiple draft tokens per request.
  * Introduced JSON schema validation during evaluation, providing grammar accuracy metrics in addition to standard accuracy.

* **Improvements**
  * Enhanced guided decoding logic for better handling of draft tokens, bitmask management, and logits processing.
  * Updated evaluation process to report both overall and grammar-specific accuracy.
  * Added tracking of draft token allocation and sequence slot information in request handling.
  * Initialized last tokens in batch requests to the last input token for improved state handling.
  * Integrated guided decoding steps into the speculative drafting workflow for improved draft token generation.
  * Improved logging verbosity and added memory monitoring stage for guided decoder initialization.
  * Refined test coverage with new guided decoding tests and updated test exclusions.
  * Allowed guided decoding with Eagle3 two-model speculative decoding and passed guided decoder to drafter creation.
  * Enhanced request and draft token management with new rollback methods for rejected and draft tokens.
  * Adjusted feature compatibility matrix to reflect guided decoding support with Eagle3 two-model engine.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by hudson-ai** [comment]
@syuoni do you have any plans to support fast-forwarded tokens in this feat (or another in the near-future)? The constraint engine can often efficiently compute a sequence of tokens that must be consumed -- you may be able to leverage the same machinery used for accepting a sequence of tokens from a draft model. We've found that the speedups from these fast-forwarded tokens can be quite substantial

**Comment by syuoni** [comment]
> @syuoni do you have any plans to support [fast-forwarded tokens](https://github.com/guidance-ai/llguidance/blob/main/docs/fast_forward.md) in this feat (or another in the near-future)? The constraint engine can often efficiently compute a sequence of tokens that must be consumed -- you may be able to leverage the same machinery used for accepting a sequence of tokens from a draft model. We've found that the speedups from these fast-forwarded tokens can be quite substantial

Hi @hudson-ai , thanks for the reminder! Actually, this feature was planned when TRT-LLM initially introduced guided decoding, namely in Nov. 2024. But my bandwidth was shifted to other stuff so I didn't have a chance to implement it.

Currently, the top priority is making guided decoding work with other features:
* Overlap scheduler (Done)
* Speculative decoding (In progress)
* Disagg serving (TODO)

I will take a closer look at the fast-forward tokens once the above work is done.

It's quite interesting to me -- on average, how many fast-forward tokens can the grammar engine generate (on any typical dataset)?

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 1]
```diff
@@ -5,91 +5,205 @@
...
+                # The last new token must be acceptable unless the matcher is terminated.
+                if matcher.is_terminated():
+                    continue
+                # TODO: Fix this.
+                last_new_token = llm_req.get_tokens(0)[-1]
+                accepted = matcher.accept_token(last_new_token)
+                # TODO: Make this an error response.
+                if not accepted:
+                    raise ValueError(
+                        f"Failed to accept last new token: {last_new_token}.")
```
Comment (syuoni) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:109]: LLGuidance + EAGLE3 fails here.

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py [hunk 2]
```diff
@@ -326,21 +327,24 @@ def create_py_executor(
...
+
     with mem_monitor.observe_creation_stage(_ExecutorCreationStage.SAMPLER):
         sampler = instantiate_sampler(model_engine, executor_config,
                                       pytorch_backend_config, mapping)
 
-    guided_decoder: Optional[GuidedDecoder] = None
-    if executor_config.guided_decoding_config is not None:
-        if spec_config is not None:
-            raise ValueError(
-                "Guided decoding is not supported with speculative decoding.")
```
Comment (syuoni) [inline tensorrt_llm/_torch/pyexecutor/py_executor_creator.py:337]: Keep raising error for one-model path.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 3]
```diff
@@ -5,91 +5,205 @@
...
 
         if len(batched_logits) > 0:
             torch.ops.trtllm.logits_bitmask(batched_logits, batched_bitmask)
+
+        if d2t is not None:
+            torch.index_select(logits, -1, d2t_mapping, out=draft_logits)
+
+    @nvtx_range("GuidedDecoder.rollback_rejected_tokens")
+    def rollback_rejected_tokens(self,
+                                 scheduled_requests: ScheduledRequests) -> None:
```
Comment (syuoni) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:177]: Add comments to explain where `rollback` methods should be called.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 4]
```diff
@@ -5,91 +5,227 @@
...
+                    matcher.fill_next_token_bitmask(self.bitmask_host[slot], i)
+                    self.num_guided_tokens[slot] += 1
+
+            if llm_req.py_is_draft:
+                assert len(llm_req.py_draft_tokens) == 0
+                self.num_advanced_draft_tokens[
+                    slot] += self.num_advanced_tokens[slot]
+
+            if (num_guided_tokens := self.num_guided_tokens[slot]) > 0:
+                with torch.cuda.stream(self._stream):
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:147]: Why are we doing this on another stream? I get that we sync with it when `execute` is called, but why is the default stream not sufficient?

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 5]
```diff
@@ -5,91 +5,227 @@
...
+                batched_bitmask.append(self.bitmask[slot, i])
+            offset += len(llm_req.py_draft_tokens) + 1
+
+        assert offset == logits.size(0)
 
         if len(batched_logits) > 0:
             torch.ops.trtllm.logits_bitmask(batched_logits, batched_bitmask)
+
+        if d2t is not None:
+            torch.index_select(logits, -1, d2t_mapping, out=draft_logits)
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:185]: Can you document that this function mutates the `logits` that are passed in? It's a bit subtle since `draft_logits` is an alias for the original `logits`, you gotta go further up to see that it's an alias for the input here

**Review by mikeiovine — COMMENTED** [review]
Mostly LGTM. Someone who has more familiarity with the stuff behind the `GuidedDecoder` abstraction should review `guided_decoder.py`, I don't have a ton of context there

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py (hunk 4 above)
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:154]: We use a standalone stream to H2D copy the bitmask, so that the H2D copy can overlap with the model forward.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py (hunk 5 above)
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:196]: Thanks! I added doc string to highlight that this is an in-place modification to the logits tensor. 

> It's a bit subtle since draft_logits is an alias for the original logits, you gotta go further up to see that it's an alias for the input here

Yes, you are right. This method intends to in-place modify the input logits tensor, which is assumed to be used for sampling later. Currently the alias logic might look a bit confusing, but I will fuse the index_copy and index_select to logits_bitmask kernel, which should make it clearer.

**Review by mikeiovine — APPROVED** [review]

**Review by nv-guomingz — APPROVED** [review]

**Review by yweng0828 — APPROVED** [review]

**Comment by Shixiaowei02** [comment]
Will this feature be compatible with dist-serving? @syuoni

**Comment by syuoni** [comment]
> Will this feature be compatible with dist-serving? @syuoni

Yes, I'm currently working on enabling guided dec x spec dec x disagg serving. A PR will be filed by the end of this week.

**Review by Shixiaowei02 — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 6]
```diff
@@ -281,6 +281,8 @@ def __init__(
             llm_request: Optional[
                 tensorrt_llm.bindings.internal.batch_manager.LlmRequest] = None,
             is_draft: bool = False,
+            seq_slot: Optional[int] = None,
+            target_seq_slot: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:285]: Will be added in https://github.com/NVIDIA/TensorRT-LLM/pull/6704
